### PR TITLE
FLW-1851-moonbeam-staking-revoke-tokens-confirm

### PR DESCRIPTION
**Pull request by MikhailPlekhanovGodoken**

### CODE REVIEW

File: common/src/main/res/values/strings.xml [hunk 1]
```diff
@@ -532,6 +532,7 @@ to your account to avoid remaining stake.</string>
...
     <string name="parachain_staking_story_rewards_page_1">Reward pool - a portion of the annual inflation that is set aside for collators and delegators.\nRewards for collators and their delegators are calculated at the start of every round for their work prior to the reward payout delay.\nThe calculated rewards are then paid out on a block-by-block basis. For every block, one collator will be chosen to receive their entire reward payout from the prior round, along with their delegators, until all of the rewards have been paid for that round.</string>
     <string name="parachain_staking_story_rewards_page_2">Delegator gets rewards after a reward payout delay. It’s a certain amount of rounds which must pass before staking rewards are distributed automatically to the free balance.\nReward distribution to some delegators may be stopped because of two possible reasons. A collator hasn\'t been chosen by the network for creating blocks or decided to leave the candidate pool. Another reason is getting a stake amount lower than a minimum bond of a collator.</string>
+    <string name="parachain_staking_redeem_title">Revoke tokens</string>
```
Comment (PankraSerg) [inline common/src/main/res/values/strings.xml:535]: Unlock tokens

File: feature-staking-impl/src/main/java/jp/co/soramitsu/feature_staking_impl/data/network/blockhain/calls/ExtrinsicBuilderExt.kt [hunk 2]
```diff
@@ -125,6 +125,32 @@ fun ExtrinsicBuilder.parachainScheduleRevokeDelegation(candidate: String): Extri
...
+        )
+    )
+}
+
+fun ExtrinsicBuilder.confirmRevokeDelegation(
+    candidateId: AccountId,
+    delegatorId: AccountId
+): ExtrinsicBuilder {
+    return call(
+        "Staking", "execute_delegation_request",
```
Comment (PankraSerg) [inline feature-staking-impl/src/main/java/jp/co/soramitsu/feature_staking_impl/data/network/blockhain/calls/ExtrinsicBuilderExt.kt:146]: Staking or ParachainStaking? if Parachain collapse this two fun's into one fun ExtrinsicBuilder.parachainExecuteDelegationRequest(...)

File: feature-staking-impl/src/main/java/jp/co/soramitsu/feature_staking_impl/presentation/staking/redeem/RedeemViewModel.kt [hunk 3]
```diff
@@ -99,25 +118,33 @@ class RedeemViewModel(
...
             coroutineScope = viewModelScope,
-            feeConstructor = { redeemInteractor.estimateFee(accountStakingFlow.first()) },
+            feeConstructor = { token ->
+                val amountInPlanks = token.planksFromAmount(amount)
+                val asset = assetFlow.first()
+
+                val stashState = accountStakingFlow.first()
+
+                redeemInteractor.estimateFee(stashState) {
+                    stakingScenarioInteractor.stakeLess(
```
Comment (PankraSerg) [inline feature-staking-impl/src/main/java/jp/co/soramitsu/feature_staking_impl/presentation/staking/redeem/RedeemViewModel.kt:139]: this should be unlock (execute_delegation_request) not stakeLess

**Review by PankraSerg — CHANGES_REQUESTED** [review]

File: common/src/main/res/values/strings.xml (hunk 1 above)
Comment (MikhailPlekhanovGodoken) [reply common/src/main/res/values/strings.xml]: Fixed

File: feature-staking-impl/src/main/java/jp/co/soramitsu/feature_staking_impl/scenarios/relaychain/StakingRelayChainScenarioInteractor.kt [hunk 4]
```diff
@@ -264,13 +262,8 @@ class StakingRelayChainScenarioInteractor(
         require(candidate != null) {
```
Comment (PankraSerg) [inline feature-staking-impl/src/main/java/jp/co/soramitsu/feature_staking_impl/scenarios/relaychain/StakingRelayChainScenarioInteractor.kt:262]: remove candiadate check - always null in relay chain

File: common/src/main/res/values/strings.xml [hunk 5]
```diff
@@ -532,6 +532,7 @@ to your account to avoid remaining stake.</string>
...
     <string name="parachain_staking_story_rewards_page_1">Reward pool - a portion of the annual inflation that is set aside for collators and delegators.\nRewards for collators and their delegators are calculated at the start of every round for their work prior to the reward payout delay.\nThe calculated rewards are then paid out on a block-by-block basis. For every block, one collator will be chosen to receive their entire reward payout from the prior round, along with their delegators, until all of the rewards have been paid for that round.</string>
     <string name="parachain_staking_story_rewards_page_2">Delegator gets rewards after a reward payout delay. It’s a certain amount of rounds which must pass before staking rewards are distributed automatically to the free balance.\nReward distribution to some delegators may be stopped because of two possible reasons. A collator hasn\'t been chosen by the network for creating blocks or decided to leave the candidate pool. Another reason is getting a stake amount lower than a minimum bond of a collator.</string>
+    <string name="parachain_staking_redeem_title">Unlock tokens</string>
```
Comment (PankraSerg) [inline common/src/main/res/values/strings.xml:535]: replace "redeem" with "unlock"

File: feature-staking-impl/src/main/java/jp/co/soramitsu/feature_staking_impl/presentation/staking/redeem/RedeemViewModel.kt [hunk 6]
```diff
@@ -99,25 +118,28 @@ class RedeemViewModel(
         router.back()
     }
 
-    fun originAccountClicked() = launch {
```
Comment (PankraSerg) [inline feature-staking-impl/src/main/java/jp/co/soramitsu/feature_staking_impl/presentation/staking/redeem/RedeemViewModel.kt:102]: the logic should be restored

File: feature-staking-impl/src/main/java/jp/co/soramitsu/feature_staking_impl/presentation/staking/redeem/RedeemFragment.kt [hunk 7]
```diff
@@ -44,7 +45,6 @@ class RedeemFragment : BaseFragment<RedeemViewModel>(R.layout.fragment_redeem) {
             redeemToolbar.setHomeButtonListener { viewModel.backClicked() }
             redeemConfirm.prepareForProgress(viewLifecycleOwner)
             redeemConfirm.setOnClickListener { viewModel.confirmClicked() }
-            redeemOriginAccount.setWholeClickListener { viewModel.originAccountClicked() }
```
Comment (PankraSerg) [inline feature-staking-impl/src/main/java/jp/co/soramitsu/feature_staking_impl/presentation/staking/redeem/RedeemFragment.kt:47]: restore logic for relay chain redeem

File: feature-staking-impl/src/main/java/jp/co/soramitsu/feature_staking_impl/presentation/staking/redeem/RedeemViewModel.kt [hunk 8]
```diff
@@ -58,37 +59,55 @@ class RedeemViewModel(
...
+                asset = it,
+                resourceManager = resourceManager,
+                retrieveAmount = { retrieveAmount },
+                patternId = patternId
+            )
+        }
+        .inBackground()
+        .asLiveData()
 
-    val originAddressModelLiveData = accountStakingFlow.map {
```
Comment (PankraSerg) [inline feature-staking-impl/src/main/java/jp/co/soramitsu/feature_staking_impl/presentation/staking/redeem/RedeemViewModel.kt:81]: restore logic for relay chain

File: feature-staking-impl/src/main/java/jp/co/soramitsu/feature_staking_impl/presentation/staking/redeem/RedeemViewModel.kt [hunk 9]
```diff
@@ -157,7 +186,6 @@ class RedeemViewModel(
 
             when {
                 payload.overrideFinishAction != null -> payload.overrideFinishAction.invoke(router)
-                result.requireValue().willKillStash -> router.returnToMain()
```
Comment (PankraSerg) [inline feature-staking-impl/src/main/java/jp/co/soramitsu/feature_staking_impl/presentation/staking/redeem/RedeemViewModel.kt:160]: restore this logic for relay chain

File: feature-staking-impl/src/main/java/jp/co/soramitsu/feature_staking_impl/scenarios/relaychain/StakingRelayChainScenarioInteractor.kt (hunk 4 above)
Comment (MikhailPlekhanovGodoken) [reply feature-staking-impl/src/main/java/jp/co/soramitsu/feature_staking_impl/scenarios/relaychain/StakingRelayChainScenarioInteractor.kt]: fixed

File: feature-staking-impl/src/main/java/jp/co/soramitsu/feature_staking_impl/presentation/staking/redeem/RedeemFragment.kt (hunk 7 above)
Comment (MikhailPlekhanovGodoken) [reply feature-staking-impl/src/main/java/jp/co/soramitsu/feature_staking_impl/presentation/staking/redeem/RedeemFragment.kt:47]: restored

File: feature-staking-impl/src/main/java/jp/co/soramitsu/feature_staking_impl/presentation/staking/redeem/RedeemViewModel.kt (hunk 8 above)
Comment (MikhailPlekhanovGodoken) [reply feature-staking-impl/src/main/java/jp/co/soramitsu/feature_staking_impl/presentation/staking/redeem/RedeemViewModel.kt:81]: restored

File: feature-staking-impl/src/main/java/jp/co/soramitsu/feature_staking_impl/presentation/staking/redeem/RedeemViewModel.kt (hunk 9 above)
Comment (MikhailPlekhanovGodoken) [reply feature-staking-impl/src/main/java/jp/co/soramitsu/feature_staking_impl/presentation/staking/redeem/RedeemViewModel.kt]: restored

File: feature-staking-impl/src/main/java/jp/co/soramitsu/feature_staking_impl/presentation/staking/redeem/RedeemViewModel.kt (hunk 6 above)
Comment (MikhailPlekhanovGodoken) [reply feature-staking-impl/src/main/java/jp/co/soramitsu/feature_staking_impl/presentation/staking/redeem/RedeemViewModel.kt:102]: restored

File: common/src/main/res/values/strings.xml (hunk 5 above)
Comment (MikhailPlekhanovGodoken) [reply common/src/main/res/values/strings.xml]: replaced

**Pull request merged**